cmd/dstask.go

```diff
@@ -1,535 +1,170 @@
 package main
 
 import (
-	"fmt"
 	"os"
-	"strconv"
-	"strings"
-	"time"
+	"path"
 
-	"github.com/mvdan/xurls"
 	"github.com/naggie/dstask"
-	"gopkg.in/yaml.v2"
 )
 
 func main() {
-	// Sets globals: GIT_REPO, STATE_FILE, IDS_FILE
-	dstask.ParseConfig()
-	dstask.EnsureRepoExists(dstask.GIT_REPO)
-	// Load state for getting and setting context
-	state := dstask.LoadState()
-	context := state.Context
+
+	repoPath := getEnv("DSTASK_GIT_REPO", os.ExpandEnv("$HOME/.dstask"))
+	dstask.EnsureRepoExists(repoPath)
+	stateFilePath := path.Join(repoPath, ".git", "dstask", "state.bin")
+	idsFilePath := path.Join(repoPath, ".git", "dstask", "ids.bin")
+	// Load state for getting and setting ctx
+	state := dstask.LoadState(stateFilePath)
+	ctx := state.Context
 	cmdLine := dstask.ParseCmdLine(os.Args[1:]...)
 
 	if cmdLine.IgnoreContext {
-		context = dstask.CmdLine{}
+		ctx = dstask.CmdLine{}
 	}
 
 	switch cmdLine.Cmd {
-	// Empty string is interpreted as CMD_NEXT
+	// The default command
 	case "", dstask.CMD_NEXT:
-		ts := dstask.LoadTasksFromDisk(dstask.NON_RESOLVED_STATUSES)
-		ts.Filter(context)
-		ts.Filter(cmdLine)
-		ts.FilterOutStatus(dstask.STATUS_TEMPLATE)
-		ts.SortByPriority()
-		context.PrintContextDescription()
-		ts.DisplayByNext(true)
-		ts.DisplayCriticalTaskWarning()
+		if err := dstask.CommandNext(repoPath, idsFilePath, stateFilePath, ctx, cmdLine); err != nil {
```

Comment:
I think it feels a bit cumbersome to pass `repoPath, idsFilePath, stateFilePath` (and perhaps even the others) for every command. I saw your comment on why you added them -- I suppose the alternatives are:

1.  Use global variables like before
2. Keep this as part of the taskset (as discussed)

What do you think?

Reply:
N.B. Otherwise I like the simplification affect this is having on dstask.go.

Reply:
There is another alternative that could reduce the number of parameters we take to the command functions in commands.go

We could consolidate these parameters into an application-level config struct.

```go
// config.go
type Config struct {
  Repo string `yaml:"repo"`
  IDsFile string `yaml:"ids_file"`
  StateFile string `yaml:"state_file"`
}
```

We then instantiate this on startup, using defaults as we do now so no one has to configure anything (unless they want to).

The config can be passed to the top level functions

```go
conf, _ := dstask.NewConfig(...)
if err := dstask.CommandNext(conf, ctx, cmdLine); if err != nil { /**/ }
```

For now, we could derive this struct from environment variables. But it lays the groundwork for a real config file, which I believe we will need eventually.

Of course, the lower level functions will look the same, as we will be passing values off this struct to the git commands, for instance.

Reply:
That seems like a good compromise 👍

Reply:
I'll @ you again when the changes are up.